src/rust/engine/rule_graph/src/lib.rs

```diff
@@ -786,30 +793,56 @@ pub fn params_str<T: TypeId>(params: &ParamTypes<T>) -> String {
   T::display(params.iter().cloned())
 }
 
-///
-/// TODO: Move all of these methods to Display impls.
-///
 pub fn entry_str<R: Rule>(entry: &Entry<R>) -> String {
-  match entry {
-    Entry::WithDeps(ref e) => entry_with_deps_str(e),
-    Entry::Param(type_id) => format!("Param({})", type_id),
+  entry_node_str_with_attrs(entry).entry_str
+}
+
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq)]
+pub struct GraphVizEntryWithAttrs {
+  entry_str: String,
+  attrs_str: Option<String>,
+}
+
+///
+/// Apply coloration to several nodes. https://c.eev.ee/kouyou/ is wonderful for selecting lovely
+/// color juxtapositions across multiple different color axes.
+pub fn entry_node_str_with_attrs<R: Rule>(entry: &Entry<R>) -> GraphVizEntryWithAttrs {
+  let (entry_str, attrs_str) = match entry {
+    &Entry::WithDeps(ref e) => (
+      entry_with_deps_str(e),
+      // Color "singleton" entries (with no params) as a nice olive green!
+      if e.params().is_empty() {
+        Some("[color=\"0.2214,0.7179,0.8528\",style=filled]".to_string())
```

Comment:
Would need to test first, but I think this could be replaced with "olive" as graphviz seems to support the svg color scheme: https://www.graphviz.org/doc/info/colors.html.
This would improve readability imo compared to float based rgb. Alternatively, if you want something very specific, I find hex representation to be a bit more readable.

From reading graphviz's doc (https://www.graphviz.org/pdf/dotguide.pdf), I see that these three floating points represent hue, saturation and brightness; but that seems pretty unusual to me as the hue is generally an int form 0-360 degrees.

Pro-tip for finding what a color looks like from its name from the command line (assuming your terminal supports 256-bit colors, but of course it does, amirite? :smile: )
```
cargo install pastel
pastel color olive
```

Reply:
If this doesn't work out, storing the colors as constants like `let olive = "0.576,0,0.6242"` would be great.

Reply:
I've added a `Palette` enum to support named colors. I kept the use of HSV for selecting these particular colors since I liked these colors specifically (in an attempt to compromise between brightness and not blocking out the graph node text), but I've left a comment next to the method that produces the `color` string for each `Palette` instance that any other color scheme can also be used to specify colors. I've added a link to https://www.graphviz.org/doc/info/colors.html in a comment as well.

Reply:
(and thank you for all the context!!!!!)